### ISSUE TITLE
SelectOneMenu: issue with uppercase characters when selecting via keyboard input

### ISSUE THREAD

**Issue by rudgal** [primefaces/primefaces#6702]
**Describe the defect**
Just recently updated to PrimeFaces 8.0 to resolve the issue described by [#4682](https://github.com/primefaces/primefaces/issues/4682). It appears that the selection of options via keyboard-input is still not working 100% as expected with respect to upper-case characters.

**Reproducer**
Reproduced with latest showcase from https://www.primefaces.org/showcase/ui/input/oneMenu.xhtml
 - PF Version: 8.0.2-SNAPSHOT
 - Affected browsers: Reproduced with Chrome Version 81.0.4044.138, also others

**Steps To Reproduce**
1. Open https://www.primefaces.org/showcase/ui/input/oneMenu.xhtml
2. Navigate via Tab-Key to field **Grouping** so that it is focused
3. Quickly type **B** **m** **w** (watch upper case for first letter)
4. Selected value is **Mercedes**

**Expected behavior**
I would expect that in described case also the value **BMW** is selected. This is already working fine when typing only lower case letters into the field. As soon as the input holds an upper-case letter, regardless of the position, it doesn't seem to work.
To be sure that this expectation also complies to the behavior with default html-selects/options I made a comparison [here](https://jsfiddle.net/rudi1024/8fs5Lrje/4). At least with the given Chrome version upper-case letters are considered correctly and the expected option is selected.

**Comment by melloware** [comment]
I will investigate.

**Comment by melloware** [comment]
@rudgal can you add this script to your application JS and reload make sure to press CTRL+F5 to reload JS code.
```javascript
if (PrimeFaces.widget.SelectOneMenu) {
    PrimeFaces.widget.SelectOneMenu.prototype.bindKeyEvents = function() {
        var $this = this;

        this.focusInput.on('keydown.ui-selectonemenu', function(e) {
                var keyCode = $.ui.keyCode,
                    key = e.which;

                switch (key) {
                    case keyCode.UP:
                    case keyCode.LEFT:
                        $this.callHandleMethod($this.highlightPrev, e);
                        break;

                    case keyCode.DOWN:
                    case keyCode.RIGHT:
                        $this.callHandleMethod($this.highlightNext, e);
                        break;

                    case keyCode.ENTER:
                        $this.handleEnterKey(e);
                        break;

                    case keyCode.TAB:
                        $this.handleTabKey();
                        break;

                    case keyCode.ESCAPE:
                        $this.handleEscapeKey(e);
                        break;

                    case keyCode.SPACE:
                        $this.handleSpaceKey(e);
                        break;
                }
            })
            .on('keyup.ui-selectonemenu', function(e) {
                var keyCode = $.ui.keyCode,
                    key = e.which;

                switch (key) {
                    case keyCode.UP:
                    case keyCode.LEFT:
                    case keyCode.DOWN:
                    case keyCode.RIGHT:
                    case keyCode.ENTER:
                    case keyCode.TAB:
                    case keyCode.ESCAPE:
                    case keyCode.SPACE:
                    case keyCode.HOME:
                    case keyCode.PAGE_DOWN:
                    case keyCode.PAGE_UP:
                    case keyCode.END:
                    case keyCode.DELETE:
                    case 16: //shift
                    case 17: //keyCode.CONTROL:
                    case 18: //keyCode.ALT:
                    case 19: //Pause/Break:
                    case 20: //capslock:
                    case 44: //Print Screen:
                    case 45: //Insert:
                    case 91: //left window or cmd:
                    case 92: //right window:
                    case 93: //right cmd:
                    case 144: //num lock:
                    case 145: //scroll lock:
                        break;

                    default:
                        //function keys (F1,F2 etc.)
                        if (key >= 112 && key <= 123) {
                            break;
                        }

                        var matchedOptions = null,
                            metaKey = e.metaKey || e.ctrlKey || e.altKey;

                        if (!metaKey) {
                            clearTimeout($this.searchTimer);

                            // #4682: check for word match
                            var text = $(this).val();
                            matchedOptions = $this.matchOptions(text);
                            if (matchedOptions.length) {
                                var matchIndex = matchedOptions[0].index;
                                var highlightItem = $this.items.eq(matchIndex);
                                if ($this.panel.is(':hidden')) {
                                    $this.selectItem(highlightItem);
                                } else {
                                    $this.highlightItem(highlightItem);
                                    PrimeFaces.scrollInView($this.itemsWrapper, highlightItem);
                                }
                            } else {
                                // #4682: check for first letter match
                                text = String.fromCharCode(key).toLowerCase();
                                // find all options with the same first letter
                                matchedOptions = $this.matchOptions(text);
                                if (matchedOptions.length) {
                                    var selectedIndex = -1;

                                    // is current selection one of our matches?
                                    matchedOptions.each(function() {
                                        var option = $(this);
                                        var currentIndex = option[0].index;
                                        var currentItem = $this.items.eq(currentIndex);
                                        if (currentItem.hasClass('ui-state-highlight')) {
                                            selectedIndex = currentIndex;
                                            return false;
                                        }
                                    });

                                    matchedOptions.each(function() {
                                        var option = $(this);
                                        var currentIndex = option[0].index;
                                        var currentItem = $this.items.eq(currentIndex);

                                        // select next item after the current selection
                                        if (currentIndex > selectedIndex) {
                                            if ($this.panel.is(':hidden')) {
                                                $this.selectItem(currentItem);
                                            } else {
                                                $this.highlightItem(currentItem);
                                                PrimeFaces.scrollInView($this.itemsWrapper, currentItem);
                                            }
                                            return false;
                                        }
                                    });
                                }
                            }

                            $this.searchTimer = setTimeout(function() {
                                $this.focusInput.val('');
                            }, 1000);
                        }
                        break;
                }
            });
    };
}
```

**Comment by rudgal** [comment]
@melloware Thanks for the quick response, I just tested the JS script.
To gain expected behavior I also had to override the matchOptions function as shown below (in order to get a non case-sensitive match):
```JS
	PrimeFaces.widget.SelectOneMenu.prototype.matchOptions = function(text) {
		return this.options.filter(function() {
			var option = $(this);
			return (option.is(':not(:disabled)') && (option.text().toLowerCase().indexOf(text.toLowerCase()) === 0));
		});
	};
```

**Comment by melloware** [comment]
@rudgal thanks for testing and the enhancement!  I have updated my PR to add the `lowercase` check.  This will be fixed for 9.0.
